client/packages/common/src/ui/forms/JsonForms/JsonForm.tsx

```diff
@@ -0,0 +1,155 @@
+import React, { FC, PropsWithChildren, useEffect } from 'react';
+import {
+  DocumentRegistryFragment,
+  Typography,
+  UnhappyMan,
+} from '@openmsupply-client/common';
+import { Box, useTranslation, BasicSpinner } from '@openmsupply-client/common';
+import { JsonForms } from '@jsonforms/react';
+import {
+  JsonFormsRendererRegistryEntry,
+  JsonSchema,
+  UISchemaElement,
+} from '@jsonforms/core';
+import { materialRenderers } from '@jsonforms/material-renderers';
+import {
+  stringTester,
+  TextField,
+  selectTester,
+  Selector,
+  groupTester,
+  Group,
+  labelTester,
+  Label,
+  dateTester,
+  Date,
+  arrayTester,
+  Array,
+} from './components';
+
+export type JsonData = {
+  [key: string]: string | number | boolean | null | unknown | JsonData;
+};
+
+interface JsonFormProps {
+  data?: JsonData;
+  documentRegistry: Pick<
+    DocumentRegistryFragment,
+    'formSchemaId' | 'jsonSchema' | 'uiSchema'
+  >;
+  isError: boolean;
+  isLoading: boolean;
+  setError: (error: string | false) => void;
+  updateData: (newData: JsonData) => void;
+}
+
+interface JsonFormsComponentProps {
+  data: JsonData;
+  jsonSchema: JsonSchema;
+  uiSchema: UISchemaElement;
+  setData: (data: JsonData) => void;
+  setError: (error: string | false) => void;
+  renderers: JsonFormsRendererRegistryEntry[];
+}
+
+// Prevents Form window being loaded with the same scroll position as its parent
+const ScrollFix = () => {
+  useEffect(() => {
+    document.getElementById('document-display')?.scrollIntoView();
+  }, []);
+  return null;
+};
+
+const FormComponent = ({
+  data,
+  jsonSchema,
+  uiSchema,
+  setData,
+  setError,
+  renderers,
+}: JsonFormsComponentProps) => {
+  return (
+    <JsonForms
+      schema={jsonSchema}
+      uischema={uiSchema}
+      data={data}
+      renderers={renderers}
+      // cells={materialCells}
+      onChange={({ errors, data }) => {
+        setData(data);
+        if (errors && errors.length) {
+          setError(errors?.map(({ message }) => message ?? '').join(', '));
+          console.warn('Errors: ', errors);
+        } else {
+          setError(false);
+        }
+      }}
+    />
+  );
+};
+
+const renderers = [
+  // We should be able to remove materialRenderers once we are sure we have custom components to cover all cases.
+  ...materialRenderers,
+  { tester: stringTester, renderer: TextField },
+  { tester: selectTester, renderer: Selector },
+  { tester: groupTester, renderer: Group },
+  { tester: labelTester, renderer: Label },
+  { tester: dateTester, renderer: Date },
+  { tester: arrayTester, renderer: Array },
+];
+
+export const JsonForm: FC<PropsWithChildren<JsonFormProps>> = ({
+  children,
+  data,
+  documentRegistry,
+  isError,
+  isLoading,
+  setError,
+  updateData,
+}) => {
+  const t = useTranslation('common');
+
+  if (isError)
```

Comment:
most of the file is just content moved from the hook. This error handler is the new part